### PR TITLE
enable users to `allow` or `deny` the security rules

### LINKED ISSUE #3877: SecurityRules should allow user to specify allow/deny a rule.
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
- While auditing SecurityRules conversion at Azure/scope, I observed that CAPZ is defaulting on `allow`-ing a specific rule. Refer to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/6232f3119fd1248580e7d14134844bb5c9af028c/azure/converters/rules.go#L35
- I wonder if the user should be allowed to specify the "Action" of a SecurityRule. 
  - If yes, we should probably update https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/6232f3119fd1248580e7d14134844bb5c9af028c/api/v1beta1/types.go#L274-L300 and other related code to allow user to specify "Action" as well as mentioned in the Azure docs on [SecurityGroups](https://learn.microsoft.com/en-us/azure/virtual-network/network-security-groups-overview#security-rules).
   
**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
- I may be missing the context on the design decision to `allow` it by default https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/6232f3119fd1248580e7d14134844bb5c9af028c/azure/converters/rules.go#L35. 
  - If this is an expected behavior and we want to have allow the custom rules specified I will close this issue with the reason specified.

**Environment:**

- cluster-api-provider-azure version: `v1.10.2`
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- This PR will let users specify `action` on the securityRules.
  - Users can specify `"Allow"` or `"Deny"` values to the security rule.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3877 

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Users can add `Deny` security rules. Security Rules are defaulted to "Allow" if unspecified.
```

### CODE REVIEW

**Comment by nawazkh** [comment]
I will squash the commits once reviewed :)

File: api/v1beta1/types.go [hunk 1]
```diff
@@ -297,6 +308,9 @@ type SecurityRule struct {
 	// Destination is the destination address prefix. CIDR or destination IP range. Asterix '*' can also be used to match all source IPs. Default tags such as 'VirtualNetwork', 'AzureLoadBalancer' and 'Internet' can also be used.
 	// +optional
 	Destination *string `json:"destination,omitempty"`
+	// Action specifies whether network traffic is allowed or denied. "Allow" or "Deny".
```
Comment (willie-yao) [inline api/v1beta1/types.go:311]: nit
```suggestion
	// Action specifies whether network traffic is allowed or denied. Can be either "Allow" or "Deny".
```

File: azure/converters/rules.go [hunk 2]
```diff
@@ -32,7 +32,7 @@ func SecurityRuleToSDK(rule infrav1.SecurityRule) network.SecurityRule {
 			SourcePortRange:          rule.SourcePorts,
 			DestinationAddressPrefix: rule.Destination,
 			DestinationPortRange:     rule.DestinationPorts,
-			Access:                   network.SecurityRuleAccessAllow,
+			Access:                   network.SecurityRuleAccess(rule.Action),
```
Comment (willie-yao) [inline azure/converters/rules.go:35]: Does this assume that `rule.Action` will always be set and is a required field? I don't see any webhook defaulting or defaulting via kubebuilder so I was just wondering.

Comment (CecileRobertMichon) [reply azure/converters/rules.go:35]: +1, in order for this to be a non-breaking change we could make the field optional and default to `Allow`

Comment (nawazkh) [reply azure/converters/rules.go:35]: +1 @willie-yao , good find! Yep. I should be defaulting Action to Allow.
I will add ` +kubebuilder:default=Basic` to the types and also make the field optional.

File: api/v1beta1/types.go (hunk 1 above)
Comment (nawazkh) [reply api/v1beta1/types.go]: Updated via 21cedcd497b5ce604a52fb2935c8a73d4dcb5607

File: azure/converters/rules.go (hunk 2 above)
Comment (nawazkh) [reply azure/converters/rules.go:35]: Updated via 21cedcd497b5ce604a52fb2935c8a73d4dcb5607

**Review by willie-yao — APPROVED** [review]
/lgtm

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Comment by nawazkh** [comment]
Sorry about the last minute hold. I wanted to update the CRDs and release note section that we are defaulting `SecurityRule.Action=allow` if unspecified.
/unhold

**Comment by nawazkh** [comment]
> pull-cluster-api-provider-azure-test — Pod got deleted unexpectedly 

🤔

**Comment by CecileRobertMichon** [comment]
/retest

likely a prow flake

**Pull request merged**